pkg/syncer/fullsync.go

```diff
@@ -468,14 +473,69 @@ func volumeInfoCRFullSync(ctx context.Context, metadataSyncer *metadataSyncInfor
 					volumeInfoCrDeletionMap[vc][cnsvolumeinfo.Spec.VolumeID] = true
 				}
 			}
+			// validate aggregated snapshot size: compare aggregated snapshot size of individual volume
+			// in cns with the size in cnsvolumeinfo. if found discrepancy in order to correct the values
+			// update the cnsvolumeinfo.
+			if isStorageQuotaM2FSSEnabled {
+				if cnsVol, ok := cnsVolumeMap[cnsvolumeinfo.Spec.VolumeID]; ok {
+					err := validateAndCorrectVolumeInfoSnapshotDetails(ctx, cnsvolumeinfo, cnsVol)
+					if err != nil {
+						log.Errorf("FullSync for VC %s: failed to update snapshot details to VolumeInfo CR "+
+							"for volume %s. Error: %+v", vc, cnsvolumeinfo.Spec.VolumeID, err)
+						continue
+					}
+				}
+			}
 		}
 	}
 	log.Debugf("FullSync for VC %s: volumeInfoCrDeletionMap: %v", vc, volumeInfoCrDeletionMap)
 }
 
+// updateVolumeInfoWithSnapshotDetails update cnsvolumeinfo with snapshotdetails by comparing
+// the aggregatedSnapshotSize of volume.
+func validateAndCorrectVolumeInfoSnapshotDetails(ctx context.Context,
+	cnsvolumeinfo *cnsvolumeinfov1alpha1.CNSVolumeInfo,
+	cnsVol cnstypes.CnsVolume) error {
+	log := logger.GetLogger(ctx)
+	log.Info("Validate Snapshot details for volumeID %q", cnsvolumeinfo.Spec.VolumeID)
+	var aggregatedSnapshotCapacity int64
+	val, ok := cnsVol.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails)
+	if ok {
+		aggregatedSnapshotCapacity = val.AggregatedSnapshotCapacityInMb
+	}
+	if aggregatedSnapshotCapacity != cnsvolumeinfo.Spec.AggregatedSnapshotSize.Value() {
+		// use current time as snapshot completion time is not available in fullsync.
+		currentTime := time.Now()
+		cnsSnapInfo := &volumes.CnsSnapshotInfo{
+			SourceVolumeID:                      cnsvolumeinfo.Spec.VolumeID,
+			SnapshotLatestOperationCompleteTime: time.Now(),
+			AggregatedSnapshotCapacityInMb:      aggregatedSnapshotCapacity,
+		}
+		log.Infof("unable to get snapshot operation completion time for volumeID %d "+
+			"will use current time %v instead", cnsvolumeinfo.Spec.VolumeID, currentTime)
+		patch := common.GetValidatedCNSVolumeInfoPatch(ctx, cnsSnapInfo, cnsvolumeinfo)
+		patchBytes, err := json.Marshal(patch)
+		if err != nil {
+			log.Errorf("error while create patch for CNSVolumeInfo instance for volume %q. Error: %+v",
+				cnsvolumeinfo.Spec.VolumeID, err)
+			return err
+		}
+		err = volumeInfoService.PatchVolumeInfo(ctx, cnsvolumeinfo.Spec.VolumeID, patchBytes,
+			allowedRetriesToPatchCNSVolumeInfo)
```

Comment:
What will happen if patch fails after max retries?

Reply:
here in case of max retries reached and there is error patching, it will return error. we will log it and move to next volume.

Reply:
@deepakkinni This seems to be a problem.  Do we have a way to get this reconciled?

Reply:
It should be okay. The next full sync will take care of it. The full sync logic takes things from CNS Query results and updates the CNSVolumeInfo to the latest aggregated snapshot size. The worst case is a patching problem, we move on, and the next full sync updates it.

Reply:
Sounds good.